rplidar.go

```diff
@@ -40,19 +45,18 @@ var (
 type Rplidar struct {
 	resource.Named
 	resource.AlwaysRebuild
-	mu                      sync.Mutex
-	close                   bool
-	device                  rplidarDevice
-	nodes                   gen.Rplidar_response_measurement_node_hq_t
-	nodeSize                int
-	started                 bool
-	scannedOnce             bool
-	defaultNumScans         int
-	warmupNumDiscardedScans int
-
-	minRangeMM float64
-
-	logger logging.Logger
+
+	deviceMutex *sync.Mutex
+	device      rplidarDevice
+	nodes       gen.Rplidar_response_measurement_node_hq_t
+	minRangeMM  float64
+
+	cancelFunc       func()
+	cacheMutex       *sync.RWMutex
+	cachedPointCloud pointcloud.PointCloud
+
+	cacheBackgroundWorkers *sync.WaitGroup
```

Comment:
Happy to make it a separate struct, will start on it now. What is the rational for not including two mutexs in the same struct? Is it simply best practice to help inform what that mutex is affecting?

Reply:
Looking at the currently rplidar struct, I think it might actually be between to include the other mutex in the device. Although this does mean storing a pointer to the rplidarDevice.

Reply:
Will look again soon - but to answer this comment, two mutexes in the same struct is totally okay - but with caution. The thing is - you're asking for memory to lock and protect the same variables within that struct, so you have to be very deliberate in the order you lock and protect them, and how you handle them across APIs. Since caching and running a background routine is a bit of a gotcha when people add code - it may make it easier to handle two mutexes protecting two structs that are accessed independently.

A good, but dumb, rule of thumb is - to make sure your mutexes are locked and then released in the same order, but this is not always the case, particularly when you have APIs calling from client code and wherever, here's a nice stack exchange I found when researching the subject - https://stackoverflow.com/questions/54681678/multiple-mutexes-in-same-struct.  Also addresses that comment I asked on why we need a pointer to a mutex at all.

There ar ea few people we work with who are excellent resources for all things concurrency and thread safety, I can point you to people far more knowledgeable than me if you want to delve into the subject more!